src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerPrimary.cs

```diff
@@ -179,5 +184,55 @@ private struct FILE_COMPLETION_INFORMATION
                 }, this).ConfigureAwait(false);
             }
         }
+
+        private class PipeReadContext
+        {
+            private readonly ListenerPrimary _listener;
+            private readonly IntPtr _bufPtr;
+            private GCHandle _bufHandle;
+            private int _bytesRead;
+
+            public PipeReadContext(ListenerPrimary listener)
+            {
+                _listener = listener;
+                _bufHandle = GCHandle.Alloc(new byte[8], GCHandleType.Pinned);
+                _bufPtr = _bufHandle.AddrOfPinnedObject();
+            }
+
+            public Libuv.uv_buf_t AllocCallback(UvStreamHandle dispatchPipe, int suggestedSize)
+            {
+                return dispatchPipe.Libuv.buf_init(_bufPtr + _bytesRead, 8 - _bytesRead);
```

Comment:
>  I guess the read callback will fail first.

Exactly. We will either call uv_close or uv_read_stop once 8 bytes are read preventing the alloc callback from being called again. Even if we didn't, we would simply return a zero length buffer which would cause the subsequent read callback to fail with the [UV_ENOBUFS status](http://docs.libuv.org/en/v1.x/handle.html#c.uv_alloc_cb).